File: screen/ansi.go
package screen

import (
	"fmt"
)


const defaultSpot =  "\x1b[1mX";

type Ansi struct {
	buffer string;
	spot string;
}

func NewAnsi() *Ansi { s:= &Ansi{buffer: "", spot: defaultSpot}; return s; }

func (screen *Ansi) SetText(x, y int, color *AColor, text string) {
	var r,g,b byte;
	r=127;g=127;b=127;
	if (color != nil) {
		r = color[R]; g = color[G]; b = color[B];
	}
	screen.buffer += fmt.Sprintf("\x1b[%d;%dH\x1b[38;2;%d;%d;%dm%s\x1b[0m", y, x, r, g, b, text);
}

func (screen *Ansi) SpotXY(x, y int, color *AColor) {
	var r,g,b byte;
	r=127;g=127;b=127;
	if (color != nil) {
		r = color[R]; g = color[G]; b = color[B];
	}
	screen.buffer += fmt.Sprintf("\x1b[%d;%dH\x1b[38;2;%d;%d;%dm%s\x1b[0m", y, x, r, g, b, screen.spot);
}

func (screen *Ansi) Clear() {
	screen.buffer = "\x1b[2J";
}

func (screen *Ansi) Flush() {
	fmt.Print(screen.buffer);
}

func (screen *Ansi) SetSpot(spot string) { screen.spot = spot; }
func (screen *Ansi) GetSpot() string { return screen.spot }

File: screen/ansi.cpp
#include "color.h"
#include "screen.h"
#include "ansi.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


Lourah::screen::Ansi::Ansi() {
	buffer = (char *)malloc(0x100000);
	spot = "\x1b[1mX";
	pos = 0;
}

Lourah::screen::Ansi::~Ansi() {
	free(buffer);
}

void Lourah::screen::Ansi::clear() {
	char const *sClear = "\x1b[2J";
	strcpy(&buffer[pos], sClear);
	pos += strlen(sClear);
}

void Lourah::screen::Ansi::flush(bool bFlush) {
	printf("%s", buffer);
	if (bFlush) fflush(stdout);
	pos = 0; // forget to reset to 0 was BIG bug !!!
}

void Lourah::screen::Ansi::spotXY(int x, int y, Lourah::screen::AColor& color) {
	pos += sprintf(&buffer[pos], "\x1b[%d;%dH\x1b[38;2;%d;%d;%dm%s\x1b[0m", y, x, color[Lourah::screen::AColor::R], color[Lourah::screen::AColor::G], color[Lourah::screen::AColor::B], spot.c_str());
}

void Lourah::screen::Ansi::setText(int x, int y, Lourah::screen::AColor&color, std::string text) {
	pos += sprintf(&buffer[pos], "\x1b[%d;%dH\x1b[38;2;%d;%d;%dm%s\x1b[0m", y, x, color[Lourah::screen::AColor::R], color[Lourah::screen::AColor::G], color[Lourah::screen::AColor::B], text.c_str());
}
